Interactive TUX

mxsdev opened the pull request:
Solves #100 

<p align="center">
<img src="https://user-images.githubusercontent.com/16108792/213215341-006bfca5-6065-468e-af3d-1e539b84f8b3.gif"/>
</p>

mxsdev commented:
@sourishkrout Oh yeah, sorry, I misunderstood the original ticket. Fixed in [4c33c94](https://github.com/stateful/runme/pull/103/commits/4c33c94bea59a33afd1021438b938bbf0974f222) and I added an `--exit` flag for the case where you want to run a command one-off.

sourishkrout commented:
Hey @silas, you have a sec to check out whether or not the UX meets your expectations?

Outside of that, @mxsdev, one thing I noticed is that the list entries can exceed your terminal's height with a fully loaded README.md such as https://github.com/stateful/vscode-runme/blob/main/examples/README.md. We could track this issue in a separate ticket, though.

silas commented:
I check tried the branch locally and it works great!

The one comment I'd add is that if the section fails it drops you out of the tui and prints a kinda ugly error message.

```sh
runme

> echo-ok sh
    echo ok
  echo-fail
    echo fail (...)

ok
runme

  echo-ok
    echo ok
> echo-fail sh
    echo fail (...)

fail
failed to run command "set -e -o pipefail;echo fail;false;\n": exit status
```

For the following readme

~~~markdown
```sh
echo ok
```

```sh
echo fail
false
```
~~~

Might be worth just printing an error message with the exist status code and sending you back into the selection menu.

Might also be nice to add a newline after the run, just to create some space between the run output and the `runme` title.

mxsdev commented:
Added scrolling (number of shown entries configurable with `--entries) and recovery from error. Also made the spacing nicer between the command's stdout and the runme tui. (and I fixed the linting errors...[for real this time](https://github.com/mxsdev/runme/actions/runs/3960604421) 😅 )

<p align="center">
<img src="https://user-images.githubusercontent.com/16108792/213512754-7302b2a5-31fa-44c9-b49f-1bc664d5f595.gif" />
</p>

I didn't give any scroll indicators, so the ui is a little minimal in this respect. Definitely let me know if you think we should add more to the ui.

@silas The ugly error message should be tracked in a separate issue, since this is the same behavior as in executing `runme run <cmd>`.

silas commented:
@mxsdev looks great! Also nice spacing change, love the polish.

I created issue #106 about improving the error message.

sourishkrout commented:
Great stuff @mxsdev! Good call on spinning out #106. @adambabik could you do a review please?

adambabik commented on internal/cmd/root.go:
```diff
@@ -16,6 +16,10 @@ var (
 )
 
 func Root() *cobra.Command {
+	exitAfterRun := false
```
Having `tuiCmd` flags here while they belong to `tuiCmd` as well as interleaving `tuiCmd` with the root command might be improved a bit.

In `root.go`, I would add these lines:

```go
tuiCmd := tuiCmd()
// tuiCmd is also a default command
cmd.RunE = tuiCmd.RunE
// remember to copy flags
tuiCmd.Flags().VisitAll(func(f *pflag.Flag) {
	pflags.AddFlag(f)
})
```

And:

```go
cmd.AddCommand(tuiCmd)
```

In this way, we keep `tuiCmd` totally unaware that it is used in any different way than other commands.

mxsdev replied:
The flag copying code errors because it re-defines the flags set by `setDefaultFlags`. In the most recent PR I used the following code:

```go
	tuiCmd.Flags().VisitAll(func(f *pflag.Flag) {
		if flag := cmd.Flags().Lookup(f.Name); flag == nil {
			cmd.Flags().AddFlag(f)
		}
	})
```

The alternative would be to never run `setDefaultFlags` on the root command and let them cascade from the tui command. I felt that it was better to be explicit than implicit here, even if at the expense of verbosity. Let me know if you want me to do it the other way!

adambabik reviewed (commented):
It's really good stuff 👍 It works very well!

I have two major comments about the code:
- `tuiModel` can be improved (removal of unnecessary pointers) so that it's more aligned with how bubbletea encourages to write programs.
- `tuiCmd()` can look like any other command and you can avoid interleaving some of its fragments with the root command.

I am happy to talk in more detail on Discord if you wish.

Thanks for the PR!

mxsdev commented on internal/cmd/tui.go:
```diff
@@ -0,0 +1,276 @@
+				isInitialRun = false
+
+				prog := tea.NewProgram(model)
+				if _, err := prog.Run(); err != nil {
```
The reason I had all those pointers initially was because I didn't realize `prog.Run()` returns the model, haha. Thanks for pointing it out!

mxsdev commented:
@adambabik Should all be implemented as you said, let me know if I missed anything!

adambabik commented on internal/cmd/run.go:
```diff
@@ -41,37 +43,45 @@ func runCmd() *cobra.Command {
-				executable.DryRun(ctx, cmd.ErrOrStderr())
-				return nil
-			}
+func runBlockCmd(block *document.CodeBlock, cmd *cobra.Command, opts *runCmdOpts) error {
```
To align the interface with `newExecutable()` you can change it to:

```suggestion
func runBlock(cmd *cobra.Command, block *document.CodeBlock, opts *runCmdOpts) error {
```

adambabik commented on internal/cmd/tui.go:
```diff
@@ -0,0 +1,263 @@
+				model = newModel.(tuiModel)
+				result := model.result
+
+				if result.run != nil {
```
You can use a technique called the early return to avoid unnecessary nesting:

```go
if result.run != nil {
    break
}

if err := runBlockCmd(result.run, cmd, nil); err != nil {
   ...
}
```

adambabik reviewed (approved):
Thank you! I added a few nitpick comments, but it's good to be merged :)

adambabik commented:
@mxsdev I have one more ask - can you resolve conflicts, please?

mxsdev commented:
Resolved conflicts, but I realized there is an error when you have a README.md file with no scripts. 

I think in this case we should probably just show a message like "No scripts to run in README.md!" And then close the program. Do you want to open a separate issue for this? Or just solve it here?

sourishkrout commented:
> Resolved conflicts, but I realized there is an error when you have a README.md file with no scripts.
> 
> I think in this case we should probably just show a message like "No scripts to run in README.md!" And then close the program. Do you want to open a separate issue for this? Or just solve it here?

I'd say let's get this merged and work on the issue above in a separate PR. We can hold off cutting a stable release until that has landed.

amac4015 commented:
iam have to say maybe a little more i dont know trainning or someting is in order since i dont really have tpo muck knowalge on whats the scope of the progect

amac4015 commented:
or not its cool good work guys
